File: cpp/organism/plant/Plant.cpp
//
// Created by Filip on 12/04/2025.
//

#include "Plant.h"

Plant::Plant(Point point, std::string symbol, int strength, int initiative,
             World &world)
    : Organism(point, symbol, strength, initiative, world) {}

void Plant::action() {
  increaseAge();
  int chance = rand() % 100;
  if (chance > getSpreadProbability() * 100)
    return;

  reproduce();
}
Point Plant::getNewPosition() {
  std::vector<Point> directions = {Point(0, 1), Point(1, 0), Point(0, -1),
                                   Point(-1, 0)};

  while (!directions.empty()) {
    int randomIndex = rand() % directions.size();
    Point newPosition = position + directions[randomIndex];

    if (world.isInBounds(newPosition) && !isOccupiedBySameType(newPosition)) {
      return newPosition;
    }

    directions.erase(directions.begin() + randomIndex); // Usuń nieodpowiedni kierunek
  }

  return position; // Jeśli nie znaleziono odpowiedniej pozycji
}

bool Plant::isOccupiedBySameType(const Point &newPosition) const {
  for (const auto &organism : world.getOrganisms()) {
    if (organism != nullptr && organism->getPosition() == newPosition &&
        organism->getSymbol() == this->getSymbol()) {
      return true;
    }
  }

  for (const auto &organism : world.getOrganismsToAdd()) {
    if (organism != nullptr && organism->getPosition() == newPosition &&
        organism->getSymbol() == this->getSymbol()) {
      return true;
        }
  }
  return false;
}

bool Plant::collision(Organism &other) {
  die();
  return false;
}

File: java/src/organism/plant/Plant.java
package organism.plant;

import organism.Organism;
import world.World;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public abstract class Plant extends Organism {

    public Plant(int x, int y, String symbol, int strength, int initiative, World world) {
        super(x, y, symbol, strength, initiative, world);
    }

    @Override
    public void action() {
        increaseAge();
        if (new Random().nextDouble() > getSpreadProbability()) return;
        reproduce();
    }

    protected int[] getNewPosition() {
        int[][] directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}};
        List<int[]> possible = new ArrayList<>();
        for (int[] d : directions) {
            int nx = x + d[0], ny = y + d[1];
            if (world.isInside(nx, ny) && !isOccupiedBySameType(nx, ny)) {
                possible.add(new int[]{nx, ny});
            }
        }
        if (possible.isEmpty()) return new int[]{x, y};
        return possible.get(new Random().nextInt(possible.size()));
    }

    protected boolean isOccupiedBySameType(int nx, int ny) {
        Organism org = world.getOrganismAt(nx, ny);
        return org != null && org.getSymbol().equals(this.getSymbol());
    }

    @Override
    public boolean collision(Organism other) {
        die();
        return false;
    }

    protected abstract void reproduce();

    protected abstract double getSpreadProbability();
}